smithy-model/src/main/java/software/amazon/smithy/model/knowledge/TextInstance.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.model.knowledge;
+
+import java.util.Deque;
+import java.util.List;
+import java.util.Objects;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.traits.Trait;
+import software.amazon.smithy.utils.ListUtils;
+
+public final class TextInstance {
+    private final TextLocation locationType;
+    private final String text;
+    private final Shape shape;
+    private final Trait trait;
+    private final List<String> traitPropertyPath;
+
+    private TextInstance(final TextLocation locationType,
+            final String text,
+            final Shape shape,
+            final Trait trait,
+            final Deque<String> traitPropertyPath) {
+        this.locationType = locationType;
+        this.text = text;
+        this.shape = shape;
+        this.trait = trait;
+        this.traitPropertyPath = traitPropertyPath != null
+            ? traitPropertyPath.stream().collect(ListUtils.toUnmodifiableList())
```

Comment:
nit - this is going to create hundreds of thousands of streams for a super simple creation. Can't you do something like: `Collections.unmodifiableList(new ArrayList<>(traitPropertyPath))`

Reply:
Changed to `ListUtils.copyOf(traitPropertyPath)`